core/src/main/scala/org/bitcoins/core/protocol/ln/LnHumanReadablePart.scala

```diff
@@ -0,0 +1,94 @@
+
+import scala.util.matching.Regex
+import scala.util.{ Failure, Success, Try }
+
+sealed abstract class LnHumanReadablePart extends NetworkElement {
+  def network: LnParams
+
+  def amount: Option[LnCurrencyUnit]
+
+  def networkToString: String = network match {
+    case LnBitcoinMainNet => "lnbc"
+    case LnBitcoinTestNet => "lntb"
+    case LnBitcoinRegTest => "lnbcrt"
+  }
+
+  override def toString: String = {
+    if (amount.isDefined) { networkToString + amount.get.toEncodedString }
+    else networkToString
+  }
+
+  override def bytes: ByteVector = ByteVector(this.toString.map(n => n.toByte))
+}
+
+case class HumanReadablePart(lnParams: LnParams, amnt: Option[LnCurrencyUnit]) extends LnHumanReadablePart {
+  require(amnt.isEmpty || amnt.get.toBigInt > 0, s"Invoice amount must be greater then 0, got $amount")
+  require(amnt.isEmpty || amnt.get <= LnPolicy.maxAmountMSat, s"Invoice amount must be less than ${LnPolicy.maxAmountMSat}, got $amount")
+
+  def network: LnParams = lnParams
+
+  def amount: Option[LnCurrencyUnit] = amnt
+}
+
+case object lnbc extends LnHumanReadablePart {
+  def network: LnParams = LnBitcoinMainNet
+  def amount: Option[LnCurrencyUnit] = None
+}
+
+case object lntb extends LnHumanReadablePart {
+  def network: LnParams = LnBitcoinTestNet
+  def amount: Option[LnCurrencyUnit] = None
+}
+
+case object lnbcrt extends LnHumanReadablePart {
+  def network: LnParams = LnBitcoinRegTest
+  def amount: Option[LnCurrencyUnit] = None
+}
+
+object LnHumanReadablePart {
+
+  private val allHRPs: Vector[LnHumanReadablePart] = Vector(lnbc, lntb, lnbcrt)
+
+  private def networkFromString(prefix: String): Try[LnParams] = {
+    val hrp = allHRPs.find(_.networkToString == prefix)
+    if (hrp.isDefined) { Success(hrp.get.network) }
+    else { Failure(new IllegalArgumentException(s"Unable to find a matching network, expected 'lnbc', 'lntb, or 'lnbcrt', got: $prefix")) }
+  }
+
+  def apply(network: NetworkParameters): Option[LnHumanReadablePart] = {
+    val lnNetwork = LnParams.fromNetworkParameters(network)
+    lnNetwork.map { params => HumanReadablePart(params, None) }
+  }
+
+  def apply(network: NetworkParameters, amount: LnCurrencyUnit): Option[LnHumanReadablePart] = {
+    val lnNetwork = LnParams.fromNetworkParameters(network)
+    lnNetwork.map { params => HumanReadablePart(params, Some(amount)) }
+  }
+
+  def apply(network: LnParams): LnHumanReadablePart = {
+    HumanReadablePart(network, None)
+  }
+
+  def apply(network: LnParams, amount: LnCurrencyUnit): LnHumanReadablePart = {
+    HumanReadablePart(network, Some(amount))
+  }
+
+  def apply(input: String): Try[LnHumanReadablePart] = {
+    val networkPattern: Regex = "^[a-z]*".r //Select all of the letters, until we hit a number, as the network
+    val networkString = networkPattern.findFirstIn(input).getOrElse("")
+    val amountString = input.substring(networkString.length) //Select the remaining part of the string as the amount
```

Comment:
Very important to note here that there is an implicit assumption that the `input` string is already formatted when it is passed in. 

What I mean by this is `input` MUST stop before the separator (`1`) that is specified in Bech32. This isn't necessarily wrong, but it should be noted on a comment on the PR. 

Sometimes when parsing things, you pass in excess information. In the case of the LN Invoice it might be the ENTIRE invoice as `input` rather than just the HumanReadablePart`. If you passed in the entire invoice, this method would fail. 

Some would say the desired behavior would be if you pass in the ENTIRE invoice, it just parses the HumanReadablePart and returns it successfully. 

Would be valid: 
`lnbc2500u`

Would not be valid:
`lnbc2500u1pvjluezpp5qqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqqqsyqcyq5rqwzqfqypqdpquwpc4curk03c9wlrswe78q4eyqc7d8d0xqzpuyk0sg5g70me25alkluzd2x62aysf2pyy8edtjeevuv4p2d5p76r4zkmneet7uvyakky2zr4cusd45tftc9c5fh0nnqpnl2jfll544esqchsrny`

Reply:
This is correct. And, I agree that it would be nice to be able to pass in an entire invoice, and return an HRP. However, we are going to have to think about this a little bit. For example, using the invoice you showed, what if we simply removed the amount (which, in my understanding, is completely valid)? We then get this `lnbc1pvjluez`.

How do we distinguish `1p` as the Bech32 separator and time field, and not read it in as 1 PicoSatoshi?

Perhaps it is not the job of LnHRP to correctly deserialize a LnInvoice? What if LnInvoice has its own deserialization, where it decodes all of the invoice pieces, and correctly identifies `1pvjluez` as the Bech32 separator and time field?

If its able to correctly identify `1pvjluez`, it could correctly pass in `lnbc` to LnHRP.

Reply:
>what if we simply removed the amount

This would invalidate the checksum on the invoice -- but you do have a great point. 

>Perhaps it is not the job of LnHRP to correctly deserialize a LnInvoice? What if LnInvoice has its own deserialization, where it decodes all of the invoice pieces, and correctly identifies 1pvjluez as the Bech32 separator and time field?

Agreed.